### PR TITLE
adding support for LGBM_BoosterUpdateOneIterCustom

### PR DESCRIPTION
**Pull request by yaxxie**
As title

### CODE REVIEW

File: src/wrapper.jl [hunk 1]
```diff
@@ -465,7 +467,35 @@ function LGBM_BoosterUpdateOneIter(bst::Booster)
...
+Also, trying to run this on a booster without data will fail.
+"""
+function LGBM_BoosterUpdateOneIterCustom(bst::Booster, grads::Vector{<:AbstractFloat}, hessian::Vector{<:AbstractFloat})
+
+    if length(bst.datasets) == 0
+        throw(ErrorException("Booster does not have any training data associated"))
+    end
+    numdata = LGBM_DatasetGetNumData(first(bst.datasets))
+
+    if !(numdata == length(grads) == length(hessian))
```
Comment (StrikerRUS) [inline src/wrapper.jl:483]: Should be equal to `numdata * LGBM_BoosterNumModelPerIteration()`
https://lightgbm.readthedocs.io/en/latest/C-API.html#c.LGBM_BoosterNumModelPerIteration
Please refer to https://github.com/microsoft/LightGBM/pull/4815#issuecomment-974569240.

Comment (FatemehTahavori) [reply src/wrapper.jl:483]: @yaxxie I will wait for this before merging it

File: test/ffi/booster.jl [hunk 2]
```diff
@@ -161,6 +161,37 @@ end
...
+    labels = randn(1000)
+    numdata = length(labels)
+    dataset = LightGBM.LGBM_DatasetCreateFromMat(mymat, verbosity)
+    LightGBM.LGBM_DatasetSetField(dataset, "label", labels)
+    # default params won't allow this to learn anything from this useless data set (i.e. splitting completes)
+    booster = LightGBM.LGBM_BoosterCreate(dataset, verbosity)
+
+    finished = LightGBM.LGBM_BoosterUpdateOneIterCustom(booster, randn(numdata), rand(numdata))
+    pred1 = LightGBM.LGBM_BoosterGetPredict(booster, 0)
+    # check both types of float work
```
Comment (FatemehTahavori) [inline test/ffi/booster.jl:176]: is this randn(numdata) type of Vector{<:AbstractFloat}?

Comment (yaxxie) [reply test/ffi/booster.jl:176]: I think the test would fail with a MethodError if it was not

File: src/wrapper.jl (hunk 1 above)
Comment (yaxxie) [reply src/wrapper.jl:483]: Yep, I have some modifications to make WRT this.

File: test/ffi/booster.jl (hunk 2 above)
Comment (FatemehTahavori) [reply test/ffi/booster.jl:176]: it is typeof Vector{Float64} which is the same as the following test Float32.(randn(numdata))

Comment (yaxxie) [reply test/ffi/booster.jl:176]: Well it's different -- the following test is Float32

The API takes only Float32 data, so if the user passes Float64, it needs to be converted. I test each way independently just to verify both entry points work as expected.

**Comment by yaxxie** [comment]
@FatemehTahavori I'm going to merge this now

File: src/wrapper.jl [hunk 3]
```diff
@@ -465,7 +467,38 @@ function LGBM_BoosterUpdateOneIter(bst::Booster)
...
+
+    if length(bst.datasets) == 0
+        throw(ErrorException("Booster does not have any training data associated"))
+    end
+    numdata = LGBM_DatasetGetNumData(first(bst.datasets))
+    nummodels = LGBM_BoosterNumModelPerIteration(bst)
+
+    if !((numdata*nummodels) == length(grads) == length(hessian))
+        throw(DimensionMismatch(
+            "Gradients sizes ($(length(grads)), $(length(hessian))) don't match training data size ($numdata) * ($nummodels)"
```
Comment (FatemehTahavori) [inline src/wrapper.jl:486]: number of trees per iteration * ($nummodels)

File: test/ffi/booster.jl [hunk 4]
```diff
@@ -161,12 +161,63 @@ end
...
+    mymat = randn(numdata, 2)
+    labels = rand((1:num_class) .- 1, numdata)
+
+    dataset = LightGBM.LGBM_DatasetCreateFromMat(mymat, verbosity)
+    LightGBM.LGBM_DatasetSetField(dataset, "label", labels)
+    booster = LightGBM.LGBM_BoosterCreate(dataset, "objective=multiclass num_class=$(num_class) $verbosity")
+
+    finished = LightGBM.LGBM_BoosterUpdateOneIterCustom(booster, randn(numdata*num_class), rand(numdata*num_class))
+    pred1 = LightGBM.LGBM_BoosterGetPredict(booster, 0)
+    # check both types of float work
```
Comment (FatemehTahavori) [inline test/ffi/booster.jl:203]: you can remove this comment

File: test/ffi/booster.jl [hunk 5]
```diff
@@ -401,4 +452,30 @@ end
 
 end
 
+
+@testset "LGBM_BoosterNumModelPerIteration" begin
+
+
+    mymat = [1. 2.; 3. 4.; 5. 6.]
+    dataset = LightGBM.LGBM_DatasetCreateFromMat(mymat, verbosity)
+    v_dataset = LightGBM.LGBM_DatasetCreateFromMat(mymat .+ 1., verbosity)
```
Comment (FatemehTahavori) [inline test/ffi/booster.jl:461]: where this have been used v_dataset?

File: src/wrapper.jl (hunk 3 above)
Comment (yaxxie) [reply src/wrapper.jl:486]: Hmm? Not sure what you're expecting here, number of trees per iteration is the number of models, so the message would be like "doesnt match training data (22) * (3)" for example

File: src/wrapper.jl [hunk 6]
```diff
@@ -465,7 +467,38 @@ function LGBM_BoosterUpdateOneIter(bst::Booster)
     return is_finished[]
 end
 
-# function LGBM_BoosterUpdateOneIterCustom()
+"""
+LGBM_BoosterUpdateOneIterCustom
+Pass grads and 2nd derivatives corresponding to some custom loss function
+grads and 2nd derivatives must be same cardinality as training data
```
[the remark this answers is not in the document]
Comment (yaxxie) [reply src/wrapper.jl:473]: Fixed

File: test/ffi/booster.jl (hunk 5 above)
Comment (yaxxie) [reply test/ffi/booster.jl:461]: Thanks, removing

**Pull request merged**